scipy/signal/filter_design.py

```diff
@@ -328,22 +349,69 @@ def freqz(b, a=1, worN=None, whole=False, plot=None):
     >>> plt.show()
 
     """
-    b, a = map(atleast_1d, (b, a))
-    if whole:
-        lastpoint = 2 * pi
-    else:
-        lastpoint = pi
+    b = atleast_1d(b)
+    a = atleast_1d(a)
+    b_len = b.shape[0]
+    a_len = a.shape[0]
+
     if worN is None:
-        N = 512
-        w = numpy.linspace(0, lastpoint, N, endpoint=False)
-    elif isinstance(worN, int):
-        N = worN
-        w = numpy.linspace(0, lastpoint, N, endpoint=False)
+        worN = 512
+
+    h = None
+    try:
+        worN = operator.index(worN)
+    except TypeError:  # not int-like
+        w = atleast_1d(worN)
     else:
-        w = worN
-    w = atleast_1d(w)
-    zm1 = exp(-1j * w)
-    h = polyval(b[::-1], zm1) / polyval(a[::-1], zm1)
+        if worN <= 0:
+            raise ValueError('worN must be positive, got %s' % (worN,))
+        lastpoint = 2 * pi if whole else pi
+        w = np.linspace(0, lastpoint, worN, endpoint=False)
+        min_size = b_len  # would be max(a_len, b_len) in the general `a` case
+        if (a_len == 1 and worN >= min_size and
+                fftpack.next_fast_len(worN) == worN):
+            # if worN is fast, 2 * worN will be fast, too,
+            # so no need to check that one
+            n_fft = worN if whole else worN * 2
+            if np.isrealobj(b) and np.isrealobj(a):
+                fft_func = np.fft.rfft
+            else:
+                fft_func = fftpack.fft
+
+            # set up the trimming to get back to worN from n_fft
+            b_trim = slice(worN) if b_len > 1 else slice(None)
+            # If we ever go back to allowing a_len > 1, we'll also need this:
+            # a_trim = slice(worN) if a_len > 1 else slice(None)
+
+            b_fft = fft_func(b, n=n_fft, axis=0)[b_trim] if b_len > 1 else b
+            # If we ever go back to using a_len != 1, we can use:
+            #     a_fft = fft_func(a, n=n_fft, axis=axis)[a_trim]
+            # and divide by a_fft. But because we know a_len == 1, we just do:
+            h = b_fft / a  # don't do inplace to allow broadcasting
+            if fft_func is np.fft.rfft and whole:
+                # If we allow a_len != 1 in this branch, this conditional
+                # would need to also check if a_len == 1
+                if b_len == 1:
+                    h = np.repeat(h, worN // 2 + 1, axis=0)
+                # exclude DC and maybe Nyquist (no need to use axis_reverse
+                # here because we can build reversal with the truncation)
+                stop = -1 if n_fft % 2 == 1 else -2
+                h_flip = slice(stop, 0, -1)
+                h = np.concatenate((h, h[h_flip].conj()), axis=0)
+            elif a_len == 1 and b_len == 1:
```

Comment:
We already know `a_len` is 1.  And see my earlier comment about eliminating the need for checking `b_len == 1`.